cognite/client/data_classes/data_modeling/typed_instances.py

```diff
@@ -65,68 +77,252 @@ def __delete__(self, instance: Any) -> None:
+            return df
+
+        pd = local_import("pandas")
+        col = df.squeeze()
+        prop_df = pd.json_normalize(col.pop("properties"), max_level=2)
+        if not prop_df.empty:
+            prop_df.columns = prop_df.columns.str.removeprefix("{}.{}/{}.".format(*self.get_source().as_tuple()))
+        return pd.concat((col, prop_df.T.squeeze())).to_frame(name="value")
+
+
+class TypedInstanceApply(WritableInstanceCore[T_CogniteResource], ABC):
+    """A typed instance. This is the write version.
+
+    Args:
+        space (str): The workspace for the instance, a unique identifier for the space.
+        external_id (str): Combined with the space is the unique identifier of the instance.
+        instance_type (Literal["node", "edge"]): The type of instance.
+        existing_version (int | None): Fail the ingestion request if the instance's version is greater than or equal to this value. If no existingVersion is specified, the ingestion will always overwrite any existing data for the instance (for the specified container or instance). If existingVersion is set to 0, the upsert will behave as an insert, so it will fail the bulk if the instance already exists. If skipOnVersionConflict is set on the ingestion request, then the instance will be skipped instead of failing the ingestion request.
+    """
+
     _instance_properties: frozenset[str]
 
+    def __init__(
+        self,
+        space: str,
+        external_id: str,
+        instance_type: Literal["node", "edge"] = "node",
+        existing_version: int | None = None,
+    ) -> None:
+        validate_data_modeling_identifier(space, external_id)
+        super().__init__(space, external_id, instance_type)
+        self.existing_version = existing_version
+
     @classmethod
     @abstractmethod
     def get_source(cls) -> ViewId:
         raise NotImplementedError
 
+    def as_apply(self) -> TypedInstanceApply:
+        return self
+
+    def as_write(self) -> TypedInstanceApply:
+        return self
+
+    def dump(self, camel_case: bool = True) -> dict[str, Any]:
+        # No dumping of sources (with properties) here, that is the job of the subclasses:
+        output: dict[str, Any] = {
+            "space": self.space,
+            "externalId" if camel_case else "external_id": self.external_id,
+            "instanceType" if camel_case else "instance_type": self.instance_type,
+        }
+        if self.existing_version is not None:
+            output["existingVersion" if camel_case else "existing_version"] = self.existing_version
+        return output
+
+    @classmethod
+    def _load(cls, resource: dict, cognite_client: CogniteClient | None = None) -> Self:
+        sources = resource.pop("sources", [])
+        properties = sources[0].get("properties", {}) if sources else {}
+        return cast(Self, _load_typed_instance(cls, resource, properties, cls._instance_properties))
+
+
+class TypedNodeApply(TypedInstanceApply):
```

Comment:
This doesn't work. TypedNodeApply needs to subclass NodeApply, and same fore TypedNode and Node. Otherwise the InstanceAPI methods don't accept them. If this was just for the datetime/int thing, then we shouldn't do that.